File: blue-core/src/blue/soundObject/Comment.java
package blue.soundObject;

import blue.*;
import blue.noteProcessor.NoteProcessorChain;
import blue.plugin.SoundObjectPlugin;
import electric.xml.Element;
import java.io.Serializable;
import java.util.Map;

/**
 * Title: blue Description: an object composition environment for csound
 * Copyright: Copyright (c) 2001 Company: steven yi music
 * 
 * @author steven yi
 * @version 1.0
 */

@SoundObjectPlugin(displayName = "Comment", live=false, position = 20)
public class Comment extends AbstractSoundObject implements Serializable,
        Cloneable {

//    private static BarRenderer renderer = new CommentRenderer();

    private String commentText;

    public Comment() {
        commentText = "";
        this.setName("Comment");
    }

    public String getText() {
        return this.commentText;
    }

    public void setText(String text) {
        this.commentText = text;
    }

    public NoteProcessorChain getNoteProcessorChain() {
        return null;
    }

    public void setNoteProcessorChain(NoteProcessorChain chain) {
    }

    public float getObjectiveDuration() {
        return this.getSubjectiveDuration();
    }

    public int getTimeBehavior() {
        return SoundObject.TIME_BEHAVIOR_NOT_SUPPORTED;
    }

    public void setTimeBehavior(int timeBehavior) {
    }

    public float getRepeatPoint() {
        return -1.0f;
    }

    public void setRepeatPoint(float repeatPoint) {
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.soundObject.SoundObject#loadFromXML(electric.xml.Element)
     */
    public static SoundObject loadFromXML(Element data,
            Map<String, Object> objRefMap) throws Exception {
        Comment comment = new Comment();

        SoundObjectUtilities.initBasicFromXML(data, comment);

        comment.setText(data.getElement("commentText").getTextString());

        return comment;

    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.soundObject.SoundObject#saveAsXML()
     */
    public Element saveAsXML(Map<Object, String> objRefMap) {
        Element retVal = SoundObjectUtilities.getBasicXML(this);

        retVal.addElement("commentText").setText(this.getText());

        return retVal;
    }

    @Override
    public NoteList generateForCSD(CompileData compileData, float startTime, float endTime) {
        return null;
    }
}

File: packages/blue-data/src/sound-objects/comment.ts
/**
 * Comment — a non-generating SoundObject for score annotations.
 * Mirrors the Java Comment class.
 */
import { AbstractSoundObject } from './abstract-sound-object';
import { NoteList } from './note-list';
import { TimeContext } from '../time/time-context';
import { CompileData } from '../compile-data';
import { Element } from '../serialization/xml-reader';
import { ObjRefSaveMap, ObjRefLoadMap } from '../serialization/obj-ref-map';
import { SoundObject } from './sound-object';
import { TimeBehavior } from './time-behavior';
import { initBasicFromXML, getBasicXML } from './sound-object-utilities';

export class Comment extends AbstractSoundObject {
  private _commentText = '';

  constructor(other?: Comment) {
    super();
    this.setName('Comment');
    this._backgroundColor = 0x404040;
    if (other) {
      this.copyFrom(other);
      this._commentText = other._commentText;
    }
  }

  getText(): string {
    return this._commentText;
  }
  setText(text: string): void {
    this._commentText = text;
  }

  override getTimeBehavior(): TimeBehavior {
    return TimeBehavior.NOT_SUPPORTED;
  }

  override generateForCSD(
    _context: TimeContext,
    _compileData: CompileData,
    _startTime: number,
    _endTime: number,
  ): NoteList {
    return new NoteList();
  }

  override saveAsXML(_objRefMap?: ObjRefSaveMap): Element {
    const elem = getBasicXML(this, 'blue.soundObject.Comment');
    elem.addElement('commentText').setText(this._commentText);
    return elem;
  }

  static loadFromXML(data: Element, _objRefMap?: ObjRefLoadMap): Comment {
    const obj = new Comment();
    initBasicFromXML(obj, data);

    const text = data.getTextString('commentText');
    if (text !== null) obj._commentText = text;

    return obj;
  }

  override deepCopy(): SoundObject {
    return new Comment(this);
  }
}
